src/envs/satellites.py

```diff
@@ -0,0 +1,598 @@
+            if not some_sat_shoots_chem_y:
+                sat = sats[rng.choice(len(sats))]
+                state_dict[sat]["shoots_chem_y"] = 1.0
+            objs = [Object(f"obj{i}", self._obj_type) for i in range(num_obj)]
+            # Sample initial positions for objects, making sure to keep
+            # them far enough apart from one another and from satellites.
+            for i, obj in enumerate(objs):
+                # Assuming that the dimensions are forgiving enough that
+                # infinite loops are impossible.
+                while True:
+                    x = rng.uniform()
+                    y = rng.uniform()
+                    geom = utils.Circle(x, y, radius)
+                    # Keep only if no intersections with existing objects.
+                    if not any(geom.intersects(g) for g in collision_geoms):
+                        break
+                collision_geoms.add(geom)
+                state_dict[obj] = {
+                    "id": i,
+                    "x": x,
+                    "y": y,
+                    "has_chem_x": 0.0,
+                    "has_chem_y": 0.0
+                }
+            init_state = utils.create_state_from_dict(state_dict)
+            goal = set()
+            for sat in sats:
+                # For each satellite, choose an object for it to read, and
+                # add a goal atom based on the satellite's instrument.
+                goal_obj_for_sat = objs[rng.choice(len(objs))]
+                if self._HasCamera_holds(init_state, [sat]):
+                    goal_pred = self._CameraReadingTaken
+                elif self._HasInfrared_holds(init_state, [sat]):
+                    goal_pred = self._InfraredReadingTaken
+                elif self._HasGeiger_holds(init_state, [sat]):
+                    goal_pred = self._GeigerReadingTaken
+                goal.add(GroundAtom(goal_pred, [sat, goal_obj_for_sat]))
+            task = Task(init_state, goal)
+            tasks.append(task)
+        return tasks
+
+    def _Sees_holds(self, state: State, objects: Sequence[Object]) -> bool:
+        sat, obj = objects
+        triangle = self._get_fov_geom(state, sat)
+        sat_x = state.get(sat, "x")
+        sat_y = state.get(sat, "y")
+        obj_x = state.get(obj, "x")
+        obj_y = state.get(obj, "y")
+        # Note: we require only that the center of the object
+        # is in the view cone, ignoring the object's radius.
+        if not triangle.contains_point(obj_x, obj_y):
+            return False
+        # Now check if the line of sight is occluded by another entity.
+        dist_denom = np.sqrt((sat_x - obj_x)**2 + (sat_y - obj_y)**2)
+        for ent in state:
+            if ent in (sat, obj):
+                continue
+            ent_x = state.get(ent, "x")
+            ent_y = state.get(ent, "y")
+            dist = abs((obj_x - sat_x) * (sat_y - ent_y) - (sat_x - ent_x) *
```

Comment:
what exactly does this distance represent? is it some kind of vector projection?

Reply:
yup, distance of each object to the line segment connecting the satellite and the object it's interacting with. this is how we check for occlusions. will add a comment